openc3/lib/openc3/utilities/local_mode.rb

```diff
@@ -0,0 +1,507 @@
+              next if gems.length > 1
+
+              if gems.length == 1
+                found_gem_name = File.basename(gems[0]).split('-')[0..-2].join('-')
+                if found_gem_name == gem_name
+                  # Same gem at least - Now see if same instance
+                  if plugin_instance
+                    if old_plugin_name
+                      # And we're updating a plugin
+                      data = File.read(plugin_instance)
+                      json = JSON.parse(data, :allow_nan => true, :create_additions => true)
+                      if json["name"] == old_plugin_name
+                        # Found plugin to update
+                        found = true
+                        update_local_plugin_files(full_folder_path, plugin_file_path, plugin_hash, gem_name)
+                      end
+                    else
+                      # New install of same plugin - Leave it alone
+                    end
+                  else
+                    # No exiting instance.json, but we found the same gem
+                    # This shouldn't happen without users using this wrong
+                    # We will update
+                    found = true
+                    update_local_plugin_files(full_folder_path, plugin_file_path, plugin_hash, gem_name)
+                  end
+                end
+              end
+            end
+
+            unless found
+              # Then we will make a local version
+              # Create a folder for this plugin and add gem and plugin_instance.json
+              folder_name = gem_name
+              count = 1
+              while File.exist?("#{OPENC3_LOCAL_MODE_PATH}/#{scope}/#{folder_name}")
+                folder_name = gem_name + "-" + count.to_s
+                count += 1
+              end
+              full_folder_path = "#{OPENC3_LOCAL_MODE_PATH}/#{scope}/#{folder_name}"
+              update_local_plugin_files(full_folder_path, plugin_file_path, plugin_hash, gem_name)
+            end
+          end
+        end
+      end
+    end
+
+    def self.update_local_plugin_files(full_folder_path, plugin_file_path, plugin_hash, gem_name)
+      return if DEFAULT_PLUGINS.include?(gem_name)
+      puts "Updating local plugin files: #{full_folder_path}"
+      FileUtils.mkdir_p(full_folder_path)
+      gems, plugin_instance = scan_plugin_dir(full_folder_path)
+      gems.each do |gem|
+        File.delete(gem)
+      end
+      temp_dir = Dir.mktmpdir
+      begin
+        unless File.exists?(plugin_file_path)
+          plugin_file_path = OpenC3::GemModel.get(temp_dir, plugin_file_path)
+        end
+        File.open(File.join(full_folder_path, File.basename(plugin_file_path)), 'wb') do |file|
+          data = File.read(plugin_file_path)
+          file.write(data)
+        end
+        File.open(File.join(full_folder_path, 'plugin_instance.json'), 'wb') do |file|
+          file.write(JSON.pretty_generate(plugin_hash, :allow_nan => true))
+        end
```

Comment:
This file is what synchronizes things. I assume it's not user editable.

Reply:
It is a user controlled file, so it is editable.   The design handles that, though usually not recommended.